src/types/program.py

```diff
@@ -102,3 +104,73 @@ def __deepcopy__(self, memo):
         return type(self).from_bytes(bytes(self))
 
     EvalError = EvalError
+
+
+def _tree_hash(node: SExp, precalculated: Set[bytes32]) -> bytes32:
+    """
+    Hash values in `precalculated` are presumed to have been hashed already.
+    """
+    if node.listp():
```

Comment:
Maybe non-recursive implementation? Since this will be consensus-critical.

Reply:
yes, that's on my list. I consider it less urgent than getting the rust implementation integrated though